### PR TITLE
chore: Improve logging, add cpu state

### LINKED ISSUE #29114: [Task]: Add logging for every api endpoint and method execution time
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

Verbose logging will help us to debug downtime issues.

The verbose logging should,

1. Log each API endpoint
2. Cron job actions
3. MongoDB and Redis queries.

### PR DESCRIPTION
**Pull request by rajatagrawal**
Fixes #29114

TODO Later On : https://github.com/appsmithorg/appsmith/issues/29581

## Summary by CodeRabbit

- **New Features**
  - Enhanced performance logging with the inclusion of request IDs for better tracking.
  - Improved diagnostics with added memory and CPU usage information during requests.

- **Refactor**
  - Updated methods for starting, stopping, and error handling to incorporate new performance logging details.

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/performancelogging/PerformanceLoggingHandler.java [hunk 1]
```diff
@@ -10,42 +9,49 @@
...
-            return false;
-        } else {
-            return true;
+        ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
+        String cpuString = "";
+
+        for (Long threadID : threadMXBean.getAllThreadIds()) {
+            ThreadInfo info = threadMXBean.getThreadInfo(threadID);
+            cpuString += String.format("{ " + info.getThreadName() + ", " + info.getThreadState() + ", "
+                    + threadMXBean.getThreadCpuTime(threadID) / 1000000 + "ms },");
```
[the remark this answers is not in the document]
Comment (rajatagrawal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/performancelogging/PerformanceLoggingHandler.java:51]: Thanks!

File: app/server/appsmith-server/src/main/java/com/appsmith/server/performancelogging/PerformanceLoggingHandler.java [hunk 2]
```diff
@@ -55,32 +61,33 @@
...
-                context.toString(),
-                executionTime,
+                "\nRequest ID : {}\nError Encountered: {}\nTime Taken {} ms\nError: {}\nContext: {}\nRequest ID: {}\n{}\n",
+                localRequestID,
                 context.getName(),
-                (context.getError() != null ? context.getError().getMessage() : null));
+                executionTime,
+                (context.getError() != null ? context.getError().getMessage() : null),
+                context.toString(),
+                memoryFootprint());
```
[the remark this answers is not in the document]
Comment (rajatagrawal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/performancelogging/PerformanceLoggingHandler.java:91]: Thanks!

File: app/server/appsmith-server/src/main/java/com/appsmith/server/performancelogging/PerformanceLoggingHandler.java [hunk 3]
```diff
@@ -10,42 +9,49 @@
...
-        }
+        requestID = (requestID + 1) % 10000;
+        context.put("requestID", requestID);
+
+        log.info(
+                "\nRequest ID : {}\nExecution Started : {} \nContext : {}\n{}\n",
+                requestID,
+                context.getName(),
+                context.toString(),
+                memoryFootprint());
```
[the remark this answers is not in the document]
Comment (rajatagrawal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/performancelogging/PerformanceLoggingHandler.java:33]: We need it to be computed everytime.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/performancelogging/PerformanceLoggingHandler.java [hunk 4]
```diff
@@ -10,42 +9,49 @@
...
 import java.lang.management.ManagementFactory;
 import java.lang.management.MemoryMXBean;
+import java.lang.management.ThreadInfo;
+import java.lang.management.ThreadMXBean;
 
 @Component
 @Slf4j
 @ConditionalOnExpression("${logging.verbose.enabled}")
 class PerformanceLoggingHandler implements ObservationHandler<Observation.Context> {
+    int requestID = 0;
```
[the remark this answers is not in the document]
Comment (rajatagrawal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/performancelogging/PerformanceLoggingHandler.java:19]: I think we can ignore thread safety for now.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/performancelogging/PerformanceLoggingHandler.java [hunk 5]
```diff
@@ -10,42 +9,50 @@
...
```
Comment (sumitsum) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/performancelogging/PerformanceLoggingHandler.java:19]: Hi @rajatagrawal can you please add a comment here explaining that this is not thread safe and it has been left as such deliberately and that we may want to address this in future ? Also, can you please create a task to address thread safety here and link it in the issue description ?

**Review by sumitsum — CHANGES_REQUESTED** [review]

File: app/server/appsmith-server/src/main/java/com/appsmith/server/performancelogging/PerformanceLoggingHandler.java (hunk 5 above)
Comment (rajatagrawal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/performancelogging/PerformanceLoggingHandler.java:19]: Done

**Review by sumitsum — APPROVED** [review]

**Pull request merged**